JS Learning Area Objects module updates

wbamberg opened the pull request:
Part of https://github.com/mdn/content/issues/10337.

What's currently done here is a rewrite of the first four parts of the JS [Objects](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects) module. What's not done is any updates to the "test your skills" bits, or any of the other pages in this module. But I think they are either going to be quite minor or it will be clear what to do about them once we've decided how to present these first four parts.

The current module begins with the following four parts:

* [Object basics](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Basics): describes object literals and walks through stuff about accessing and modifying members, and a bit about `this`. Pretty inoffensive.

* [Object-oriented JavaScript for beginners](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Object-oriented_JS): starts off going through some conceptual OOP stuff, which is mostly OK though I had issues with some of it - I think headings should reflect real terminology (i.e. use "inheritance" rather than "Specialist classes") and I didn't like the way the diagrams for class/instance and for inheritance looked almost identical though they are very different things. It then talks about JS constructors, which was mostly OK but seemed like a bit of a non-sequitur.

* [Object prototypes](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Object_prototypes): this tallked about prototypes and the prototype chain, then got into how to use prototypes to implement class-based inheritance. I must say I found this extremely hard going, right from the start! I'm not sure how beginners could follow this.

* [Inheritance in JavaScript](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Inheritance): this continued where the last article left off and then half way through switched into `class` syntax and showed the (much simpler) way to do it there.

One thing I wanted to avoid was the suggestion that prototypes are basically just a crap way of doing OOP in JS, that's all fixed by `class`. The problem is if you *just* present prototypes as a way of implementing class-based OOP, then it will tend to end up like this. I think that gives a false impression, as there are things prototypes are useful for, that are not class-based OOP (for example, you can see them as a delegation system more than an inheritance system).

So at a high level I wanted to separate out discussion of prototypes in JS from discussion of implementing class-based OOP in JS. I think that *if* you want to do class-based OOP in JS, then `class` is really helpful. But class-based OOP is not at all the only way you can combine objects in JS, and sometimes (maybe often) something more lightweight is better. Although I didn't want to go into those alternatives, I at least wanted to leave that open.

So I have:

* **Object basics**: starts out like the current one but then talks about constructors. Note that we haven't talked about OOP yet, so this is just constructors in themselves, as a useful way to create object instances. I am conflicted about whether to make the constructors bit into its own article. That would make it easier to refer to, but it would be pretty short.

* **Object prototypes**: explains what prototypes are, and how the chain works, and how to set your own prototypes. Again, no OOP yet, so this is just prototypes as themselves, not as a means to implementing any system. But a bit at the end about how they might relate to inheritance.

* **Object-oriented programming**: mostly just conceptual stuff about class-based OOP: classes versus instances, inheritance, encapsulation. Then at the end I've tried to tie that to prototypes, and tried to be careful about how the prototype model is similar, how it is different, and how if you want to write class-based OOP in JS, there is this `class` stuff you should look at rather than trying to build it directly using prototypes.

* **Classes in JavaScript**: this just presents `class` and so on, and is very straightforward. I didn't cover everything here - maybe I should include static, and getters/setters?

******

At this point I'm mostly interested in general feedback about whether this is a good way to present things overall. If we agree it is, I'll update all the other bits in this module and ask for a full review. (That is, I'm happy to accept detailed feedback now, but will ask for a full review eventually anyway).

wbamberg committed:
Rewrite OO guide in JS Learning Area

wbamberg committed:
Fix a typo

wbamberg committed:
Lowercase SVG filenames

wbamberg committed:
Compress SVG

Rumyra commented on files/en-us/learn/javascript/objects/object-oriented_programming/index.md:
```diff
@@ -0,0 +1,252 @@
+  </tbody>
+</table>
+
+Object-oriented programming is about modeling a system as a collection of objects, which each represent some particular aspect of the system. Objects contain both functions (or methods) and data. An object provides a public interface to other code that wants to use it, but it maintains its own private internal state: this means that other parts of the system don't have to care about what is going ...
```
Same `;` or `T`

Rumyra commented on files/en-us/learn/javascript/objects/object-oriented_programming/index.md:
```diff
@@ -0,0 +1,252 @@
+
+## Classes and instances
+
+In OOP, when we model a problem in terms of objects we create abstract definitions representing the types of object we want to have in our system. For example, if we were modeling a school, we might want to have objects representing professors. Every professor has some properties in common: they all have a name and a subject that they teach. Also, every professor can do certain things: for exampl ...
```
same

Rumyra commented on files/en-us/learn/javascript/objects/object-oriented_programming/index.md:
````diff
@@ -0,0 +1,252 @@
+So in pseudocode a `Professor` class could be written like this:
+
+```
+class Professor
+    methods
+        grade(paper)
+        introduceSelf()
+    properties
+        name
+        subject
````
Do you think methods & properties should be the other way around? As you'd use `name` in `introduceSelf()`

wbamberg replied:
I like (public) methods-first because I think it emphasises the public interface, but you are right, this is probably clearer for people. I just pushed https://github.com/mdn/content/pull/11373/commits/a9852d232728ab47b1efa6995161d83a6d0d46f9 for this.

Rumyra commented on files/en-us/learn/javascript/objects/object-oriented_programming/index.md:
````diff
@@ -0,0 +1,252 @@
+Now we have a constructor we can create some professors. Programming languages often use the keyword `new` to signal that a constructor is being called.
+
+```
+walsh = new Professor('Walsh', Psychology')
+lillian = new Professor('Lillian', Poetry')
````
```suggestion
walsh = new Professor('Walsh', 'Psychology')
lillian = new Professor('Lillian', 'Poetry')
```

Rumyra commented on files/en-us/learn/javascript/objects/object-oriented_programming/index.md:
```diff
@@ -0,0 +1,252 @@
+walsh = new Professor('Walsh', 'Psychology')
+walsh.introduceSelf()  // 'My name is Professor Walsh, and I will be your Psychology professor'
+
+summers = new Student('Summers', '1')
```
Is `'1'` a string?

wbamberg replied:
no, good catch. Updated in https://github.com/mdn/content/commit/a9852d232728ab47b1efa6995161d83a6d0d46f9.

Rumyra reviewed (changes requested):
Looks really good @wbamberg - just a couple of nits 👍

wbamberg commented on files/en-us/learn/javascript/objects/classes_in_javascript/index.md:
```diff
@@ -0,0 +1,232 @@
+
+We also talked about how it's possible to use [prototypes](Learn/JavaScript/Objects/Object_prototypes) to implement a model like this, and that JavaScript also provides features that map more closely to classical OOP concepts.
+
+In this article we'll go through these features. It's worth keeping in mind that the features described here are not a new way of combining objects: under the hood, they still use prototypes. They're just a way to make it easier to set up a prototype chain.
```
That's interesting! I would never do this, but according to https://klwightman.com/2018/01/08/when-do-you-capitalize-after-a-colon/ and https://www.grammarly.com/blog/capitalization-after-colons/ it is required according to the APA style (but not according to the Chicago Manual of Style).

wbamberg replied:
...and checking, we apparently go with the Chicago Manual of Style: https://developer.mozilla.org/en-US/docs/MDN/Guidelines/Writing_style_guide#preferred_style_guides .

Rumyra replied:
Ha - well I learnt something today - should probably read the Chicago Manual of Style - some light holiday reading 👍

wbamberg committed:
remove extra space

Co-authored-by: Ruth John <Rumyra@users.noreply.github.com>

wbamberg committed:
Missing quotes

Co-authored-by: Ruth John <Rumyra@users.noreply.github.com>

wbamberg committed:
List properties before methods in class pseudocode

wbamberg committed:
year parameter should be a number

hamishwillee committed:
Minor change to my taste

hamishwillee commented on files/en-us/learn/javascript/objects/basics/index.md:
```diff
@@ -246,32 +241,102 @@ Adding a property to an object using the method above isn't possible with dot n
+
+Using object literals is fine when you only need to create one object, but if you have to create more than one, as in the previous section, they're seriously inadequate. We have to write out the same code for every object we create, and if we want to change some properties of the object - like adding a `height` property - then we have to remember to update every object.
+
+We would like a way to define the "shape" of an object - the set of methods and the properties it can have - and then create as many objects as we like, just updating the values for the properties that are different.
```
emdash?

```suggestion
We would like a way to define the "shape" of an object — the set of methods and the properties it can have — and then create as many objects as we like, just updating the values for the properties that are different.
```

hamishwillee commented:
> Object basics: starts out like the current one but then talks about constructors. Note that we haven't talked about OOP yet, so this is just constructors in themselves, as a useful way to create object instances. I am conflicted about whether to make the constructors bit into its own article. That would make it easier to refer to, but it would be pretty short.

This looks great to me. 

I like constructors in this doc - its a clear story. I guess whether or not you pull them out depends on where you want to refer to them from.

hamishwillee commented on files/en-us/learn/javascript/objects/object_prototypes/index.md:
```diff
@@ -42,269 +29,211 @@ Prototypes are the mechanism by which JavaScript objects inherit features from o
+We then set the `Person` function's `prototype` property to point to `personPrototype`.
 
-> **Note:** If you are having trouble getting this example to work, have a look at our [oojs-class-prototype.html](https://github.com/mdn/learning-area/blob/master/javascript/oojs/advanced/oojs-class-prototype.html) example (see it [running live](https://mdn.github.io/learning-area/javascript/oojs/advanced/oojs-class-prototype.html) also).
+The last line `Person.prototype.constructor = Person;` needed because the prototype has a `constructor` property that points to the constructor function that was used to create the object. Since we have replaced the `prototype` property, we need to set `prototype.constructor` back to its original value.
```
```suggestion
The last line (`Person.prototype.constructor = Person;`) is needed because the prototype has a `constructor` property that points to the constructor function that was used to create the object. Since we have replaced the `prototype` property, we need to set `prototype.constructor` back to its original value.
```

This bit confuses me a little:

> .. because the prototype has a `constructor` property that points to the constructor function that was used to create the object. 

So the prototype is `personPrototype` - but it is an object literal. So what constructor was used to create it? We didn't use `new` to create it and we haven't introduced that it might have one somehow under the hood. Or am I missing something? I think I get the point, but it feels a bit "inception"

wbamberg replied:
It is really confusing. I don't like this bit at all, because I want to treat it as an aside (which it is, really) but it's complicated enough to get snagged on.

AFAICT it's about the instance that's created, `reuben`, and what the result is of accessing its `constructor` property. For instance, try not setting `constructor` as we do above:

```js
const personPrototype = {
  greet() {
    console.log(`hello, my name is ${this.name}!`);
  }
}

function Person(name) {
  this.name = name;
}

Person.prototype = personPrototype;

const reuben = new Person('Reuben');

console.log(reuben.constructor);
```

This will give us `function Object()` (because the constructor for the `personPrototype` instance was `Object` (because we created it as an object literal)). This is obviously wrong - the constructor for `reuben` should be  `Person()`, because that's what we used to create the `reuben` instance.

I want to say here "don't look too deeply into this at this point, this is just a thing you have to do" but I never like saying this in docs. There is more detail here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/constructor#changing_the_constructor_of_a_function  but honestly I didn't find it very helpful.

hamishwillee commented on files/en-us/learn/javascript/objects/object_prototypes/index.md:
````diff
@@ -42,269 +29,211 @@ Prototypes are the mechanism by which JavaScript objects inherit features from o
-```js
-Person.prototype.fullName = this.name.first + ' ' + this.name.last;
-```
+### Own properties
````
FYI hasOwnProperty is supposed to be replaced eventually by [Object.hasOwn()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn). 

Might we worth mentioning here as it is better. Might not as it is not widely implemented. Perhaps a note?

wbamberg replied:
I wonder if I should also change the usage of `hasOwnProperty` earlier on?

Well, for now I'll just add a note down here.

wbamberg replied:
I just pushed https://github.com/mdn/content/pull/11373/commits/ac1ca1aca46c97249406af6d7108c624f78f0670 to talk about `hasOwn` here and add a note about `hasOwnProperty` (giving primacy to the new version, since it seems to be supported in most browsers).

hamishwillee replied:
Works for me.

hamishwillee commented on files/en-us/learn/javascript/objects/object_prototypes/index.md:
```diff
@@ -42,269 +29,211 @@ Prototypes are the mechanism by which JavaScript objects inherit features from o
+Every object in JavaScript has a built-in property, which is called its **prototype**. The prototype is itself an object, so the prototype will have its own prototype, making what's called a **prototype chain**. The chain ends when we reach a prototype that has `null` for its own prototype.
 
-2.  The output won't show you very much because we haven't defined anything on our custom constructor's prototype! By default, a constructor's `prototype` always starts empty. Now try the following:
+> **Note:** The property of an object that points to its prototype is **not** called `prototype`. Its name is not standard, but in practice all browsers use {{jsxref("Object/__proto__", "__proto__")}}. The standard way to access an object's prototype is the {{jsxref("Object/getPrototypeOf", "Object.getPrototypeOf()")}} method.
```
So to be clear:
- An object's prototype property is pointed called `/__proto__`
- A constructor function property named `prototype` will be set to the prototype of the newly created object.
- So if you construct an object from a constructor, under the hood there will be an assignment that looks like `newobject.__proto__=constructorfunction.prototype`

Is that right?

wbamberg replied:
> So to be clear:
> 
>     * An object's prototype property is pointed called `/__proto__`
> 
>     * A constructor function property named `prototype` will be set to the prototype of the newly created object.
> 
>     * So if you construct an object from a constructor, under the hood there will be an assignment that looks like `newobject.__proto__=constructorfunction.prototype`
> 
> 
> Is that right?
> 

100%.

> I think I found it jarring that we mentioned that the object prototype property is not named prototype, and then in the constructor we have a property named prototype that assigns the prototype for the created object.

The fact that the constructor has a property `prototype` is exactly why we had that note - the name of the property that references the prototype is `__proto__` not `prototype`.

> But then it occurs to me that the constructor.prototype is NOT the constructors prototype - that would be named `__proto__`.

Yes, exactly.

> I think I might be thick.

No, this stuff is very confusing.

hamishwillee replied:
Cool. I think perhaps we can help a bit by attempting to reiterate the difference between `prototype` and `__proto__` again - I have suggested this in https://github.com/mdn/content/pull/11373/files#r777761617 - for your consideration.

hamishwillee commented on files/en-us/learn/javascript/objects/object_prototypes/index.md:
````diff
@@ -42,269 +29,211 @@ Prototypes are the mechanism by which JavaScript objects inherit features from o
+This is an object with one data property, `city`, and one method, `greet()`. If you type the object's name and a period into the console, like `myObject.`, then the console will show you a list of all the properties available to this object. You'll see that as well as `city` and `greet`, there are lots of other properties!
 
-![](mdn-graphics-person-person-object-2.png)
+```
````
When I type in myObject I do see lots of other items, but they are grouped under this `[[prototype]]` thingy which I assume represents `__proto__`. Is this mismatch in what we see something that needs to be explained?

![image](https://user-images.githubusercontent.com/5368500/147722792-377fc31e-bae7-476b-a034-9f3621c4bffd.png)

wbamberg replied:
In the description:

> If you type the object's name and a period into the console, like `myObject.`, then the console will show you a list of all the properties available to this object.

...the period is important. If you just type `myObject` and return, then yes, you get the console's representation of the object. If you type `myObject.` without a return, you get the autocomplete popup, which is what I was trying to show.

I guess this is tricky! Maybe it would be better to do something like:

> This is an object with one data property, `city`, and one method, `greet()`.
>
> Now try typing this into the console:
>
> ```js
> console.log(myObject.hasOwnProperty());
> console.log(myObject.toString());                  
> ```
> You should see something like this:
>
> ```
> false
> [object Object]
> ```
> What are these extra functions, and where do they come from?

It's a bit of a shame not to show all the extra properties, but this is probably easier to follow. What do you think?

hamishwillee replied:
No, I think you've done it just right - I just missed the period because I assumed I knew how this works. I have proposed using style to make this more obvious https://github.com/mdn/content/pull/11373/files#r777753158

hamishwillee replied:
I'll mark this as resolved, up to you if you like the proposed fix.

hamishwillee commented:
> Object prototypes: explains what prototypes are, and how the chain works, and how to set your own prototypes. Again, no OOP yet, so this is just prototypes as themselves, not as a means to implementing any system. But a bit at the end about how they might relate to inheritance.

This is really good IMO - simple concepts and well untangled from classes and OO other than noting its use for inheritance.

hamishwillee commented on files/en-us/learn/javascript/objects/object-oriented_programming/index.md:
```diff
@@ -0,0 +1,255 @@
+
+Object-oriented programming is about modeling a system as a collection of objects, which each represent some particular aspect of the system. Objects contain both functions (or methods) and data. An object provides a public interface to other code that wants to use it, but it maintains its own private internal state: this means that other parts of the system don't have to care about what is going ...
+
+> **Note:** This article describes the concepts of OOP, independent of any particular programming language. To emphasise this all the examples given are given in {{Glossary("Pseudocode", "pseudocode")}}.
```
Perhaps move this note to the top below or as part of the first note. 

In addition, when I read that first section my thinking was "**why** are you telling me about these things in terms of another programming language".

I understand why - because there isn't a precise mapping and javascript has other tools and methods that allow you to do similar things. But it might be worth telling the reader.

wbamberg replied:
OK, I've had a go at this in https://github.com/mdn/content/pull/11373/commits/3a83234d0aac68c1a13c35de33bee02cf75d8508.

hamishwillee replied:
I think this is better. Thanks.

hamishwillee commented on files/en-us/learn/javascript/objects/object-oriented_programming/index.md:
```diff
@@ -0,0 +1,255 @@
+
+class Professor : extends Person
+    properties
+        subject
```
Perhaps "teaches" (everywhere). Because students take subjects too.

wbamberg replied:
Yes! -> https://github.com/mdn/content/pull/11373/commits/b8491623b54805b63bb749ce8d6adb0b36c807d5

hamishwillee commented on files/en-us/learn/javascript/objects/classes_in_javascript/index.md:
```diff
@@ -0,0 +1,232 @@
+
+In [the last article](Learn/JavaScript/Objects/Object-oriented_programming) we introduced some basic concepts of object-oriented programming (OOP), and discussed an example where we used OOP principles to model professors and students in a school.
+
+We also talked about how it's possible to use [prototypes](Learn/JavaScript/Objects/Object_prototypes) to implement a model like this, and that JavaScript also provides features that map more closely to classical OOP concepts.
```
... prorotypes _and constructors_ ?

wbamberg replied:
-> https://github.com/mdn/content/pull/11373/commits/30dd13a858cfcaca6438f42c96c4402b1edbb013

hamishwillee commented on files/en-us/learn/javascript/objects/classes_in_javascript/index.md:
```diff
@@ -0,0 +1,232 @@
+
+You could also initialize the property to a default value when you declare it, with a line like `name = '';`.
+
+The constructor is identified using the {{jsxref("Classes/constructor", "constructor")}} keyword. Just like a standalone constructor, it will:
```
add link for "standalone constructor" ?

wbamberg replied:
-> https://github.com/mdn/content/pull/11373/commits/30dd13a858cfcaca6438f42c96c4402b1edbb013

hamishwillee commented on files/en-us/learn/javascript/objects/classes_in_javascript/index.md:
````diff
@@ -0,0 +1,232 @@
+giles.introduceSelf(); // Hi! I'm Giles
+```
+
+Note that the name of the constructor is the name of the class.
````
That could be  confusing. We just said the name of the constructor function is `constructor` ...

wbamberg replied:
We didn't say it was named constructor, we said it was identified using the constructor keyword. Actually "identified" is not very good here and "defined" is probably better.

But I've made some changes in https://github.com/mdn/content/pull/11373/commits/30dd13a858cfcaca6438f42c96c4402b1edbb013 which hopefully help with this.

hamishwillee replied:
It helped me :-)

hamishwillee commented on files/en-us/learn/javascript/objects/classes_in_javascript/index.md:
```diff
@@ -0,0 +1,232 @@
+
+Since we want to set the subject when a new `Professor` is created, we define a constructor,  which takes the `name` and `subject` as arguments. The first thing this constructor does is call the superclass constructor using {{jsxref("Operators/super", "super()")}}, passing up the `name` parameter. The superclass constructor takes care of setting `name`. After that the `Professor` constructor sets ...
+
+> **Note:** If a subclass has any of its own initialization to do, it **must** first call the superclass constructor using `super()`, passing up any parameters that the superclass constructor is expecting.
```
This implies that this must be done, and it must be done "first". You can't do it at the end of the constructor? What if you do?

wbamberg replied:
You will get a runtime error when you try to construct an instance of the class:

> Uncaught ReferenceError: must call super constructor before using 'this' in derived class constructor

hamishwillee commented:
As discussed, maybe a little more "glue" at the top of the object orientation one to explain why we are doing the split between OOP and classes docs this way might be useful. Just a line or two.

I think this is a really excellent structure - much better than what it replaces. I hadn't "faded out" by the end of the review either, which is a very good sign :-).

wbamberg committed:
Use emdash

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

wbamberg committed:
Update files/en-us/learn/javascript/objects/object_prototypes/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

wbamberg committed:
Use hasOwn rather than hasOwnProperty

wbamberg committed:
subject->teaches

wbamberg committed:
Various review comments

wbamberg committed:
Clarify overview of OOP article

wbamberg commented:
> As discussed, maybe a little more "glue" at the top of the object orientation one to explain why we are doing the split between OOP and classes docs this way might be useful. Just a line or two.
> 
> I think this is a really excellent structure - much better than what it replaces. I hadn't "faded out" by the end of the review either, which is a very good sign :-).

Thanks very much for taking a look @hamishwillee !

I have tried to address everything except for the question in https://github.com/mdn/content/pull/11373#discussion_r777646216 - let me know if you think the version described in that comment is better than what we have now.

hamishwillee commented on files/en-us/learn/javascript/objects/object_prototypes/index.md:
```diff
@@ -42,269 +29,213 @@ Prototypes are the mechanism by which JavaScript objects inherit features from o
+We then set the `Person` function's `prototype` property to point to `personPrototype`.
 
-> **Note:** If you are having trouble getting this example to work, have a look at our [oojs-class-prototype.html](https://github.com/mdn/learning-area/blob/master/javascript/oojs/advanced/oojs-class-prototype.html) example (see it [running live](https://mdn.github.io/learning-area/javascript/oojs/advanced/oojs-class-prototype.html) also).
+The last line (`Person.prototype.constructor = Person;`) is needed because the prototype has a `constructor` property that points to the constructor function that was used to create the object. Since we have replaced the `prototype` property, we need to set `prototype.constructor` back to its original value.
```
Following on from https://github.com/mdn/content/pull/11373#discussion_r777624481 - this is better but still confusing. Perhaps we can make this more clear by just state what we're achieving:

```suggestion
The last line (`Person.prototype.constructor = Person;`) sets the prototype constructor property to the function used to create `Person` objects.
This is required because after setting `Person.prototype = personPrototype;` the property points to the constructor for the `personPrototype`, which is `Object` rather than `Person` (because `personPrototype` was constructed as an object literal).
```

Thanks for the explanation. I see why you want this to be an aside-generally you don't directly call this or need to call it. I did think of adding a note to that effect, but decided that at this point "just follow the rule" is better.

hamishwillee commented on files/en-us/learn/javascript/objects/object_prototypes/index.md:
```diff
@@ -42,269 +29,213 @@ Prototypes are the mechanism by which JavaScript objects inherit features from o
-![](object-available-members.png)
-
-In this list, you will see the members defined on `person1`'s constructor — `Person()` — `name`, `age`, `gender`, `interests`, `bio`, and `greeting`. You will however also see some other members — `toString`, `valueOf`, etc — these are defined on `person1`'s prototype object's prototype object, which is `Object.prototype`.
+This is an object with one data property, `city`, and one method, `greet()`. If you type the object's name and a period into the console, like `myObject.`, then the console will show you a list of all the properties available to this object. You'll see that as well as `city` and `greet`, there are lots of other properties!
```
Perhaps - following on from https://github.com/mdn/content/pull/11373/files#r776568326  - 

```suggestion
This is an object with one data property, `city`, and one method, `greet()`. If you type the object's name _followed by a period_ into the console, like `myObject.`, then the console will pop up a list of all the properties available to this object. You'll see that as well as `city` and `greet`, there are lots of other properties!
```

wbamberg committed:
Clarify instructions for auto-complete

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

hamishwillee commented on files/en-us/learn/javascript/objects/object_prototypes/index.md:
```diff
@@ -42,269 +29,213 @@ Prototypes are the mechanism by which JavaScript objects inherit features from o
+### Using a constructor
 
-You should get an alert message displayed, featuring the person's name as defined inside the constructor. This is really useful, but what is even more useful is that the whole inheritance chain has updated dynamically, automatically making this new method available on all object instances derived from the constructor.
+In JavaScript, all functions have a property named `prototype`. When you call a function as a constructor, the prototype of the object the constructor creates is set to the constructor's `prototype`.
```
This is just a suggestion, perhaps worth re-iterating here 
```suggestion
In JavaScript, all functions have a property named `prototype`. When you call a function as a constructor, this property is set as the prototype of the newly constructed object (by convention, in the property named `__proto__`).
```

We might also consider a note after:

> **Note** The terminology is confusing.  The `prototype` property is the prototype that gets assigned to objects constructed from the function. The prototype of the function itself is stored in its `__proto__` property (as for all  other objects).

hamishwillee commented:
Thanks, fixes look good to me. I added a few more suggestions.

wbamberg committed:
clarify explanation of setting prototype.constructor

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

wbamberg committed:
Update files/en-us/learn/javascript/objects/object_prototypes/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

wbamberg committed:
Update object basics test your skills

wbamberg committed:
Revise OO-JS test your skills page

wbamberg committed:
Use fetch in JSON article

wbamberg committed:
Fix intra-module links

wbamberg committed:
Update JSON Test Your Skills

wbamberg committed:
Update object building practice to use class syntax

wbamberg committed:
Update bouncing balls assessment to work with classes

wbamberg committed:
Merge remote-tracking branch 'upstream/main' into js-la-object-oriented-js

* upstream/main: (286 commits)
  MediaQueryList removeListener depreciation (#11749)
  Update index.md to make text time independent (#11748)
  Corrected typos in multiple pages (#11746)
  Fix some awkward hyperlink formatting (#11740)
  HTMLElement .offset[height|left|parent|top|width] are stable removed experimental icon (#11741)
  Typo (#11742)
  Fix comments (#11756)
  Restore line breaks in the `<dd>`s (#11757)
  Added clarification when property reflects the HTML attribute (#11751)
  Bump technote-space/get-diff-action from 5.0.2 to 6.0.1 (#11755)
  Change "validate" to "check" (#11550)
  Clarify behavior of Array.p.lastIndexOf() fromIndex arg (#11731)
  FF97 Add experimental feature for ScreenOrientation.lock() (#11705)
  Correcting one of the example provided (#11709)
  Adds support for AbortSignal.reason and related items (#11708)
  Added documentation on missing property (#11721)
  Bump @mdn/yari from 0.5.4 to 0.6.6 (#11722)
  Fix: bring gl.FRAMEBUFFER target in line with  the spec (#11732)
  ScreenOrientation.lock() updates for FF97 (#11684)
  Reword actualBoundingBoxLeft description (#11290)
  ...

wbamberg commented:
OK, I just pushed some more commits to update the other articles in this module. I think these should be more "mechanical" changes.

* https://github.com/mdn/content/pull/11373/commits/7cc4a49b9d7c9aec6b7164f0be0d23cb3739ea8a: updates [Test your skills: Object basics](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Test_your_skills:_Object_basics) to add a new task to use a constructor.
* https://github.com/mdn/content/pull/11373/commits/d1572d904075f2e36b874aa57029263aaba5471a: updates [Test your skills: Object-oriented JavaScript](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Test_your_skills:_Object-oriented_JavaScript) to test only `class` syntax. This means we remove one of the tasks (originally the first task was old-style, then task 2 was to convert it to `class` syntax, then the third adds a subclass. I've removed the first task, effectively, so we start with `class` syntax).
* https://github.com/mdn/content/pull/11373/commits/51e4e874b8bbee43a9a0455fbe9eeb6cc9e1644e: updates [Working with JSON](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/JSON) to use fetch instead of XHR, and sundry little changes like `for..of`.
* https://github.com/mdn/content/pull/11373/commits/c58a73b63e46eefbeb78e7a389a5905785f4b58f: minor changes to keep all the links between articles working. 
* https://github.com/mdn/content/pull/11373/commits/01acf66327ba3d205ae4ae01dc9fd68ad28b0768: updates [Test your skills: JSON](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Test_your_skills:_JSON). The weird thing here is that the article says you can't run the example locally because of cross-origin restrictions, but it seems to work just fine if you use an absolute URL? So I changed it. But maybe I'm wrong here?
* https://github.com/mdn/content/pull/11373/commits/a24b324f35168c9f51c4c14c525fb0291e2a60a3: updates [Object building practice](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Object_building_practice) to use `class` syntax, plus a few other minor improvements.
* https://github.com/mdn/content/pull/11373/commits/cdb5629be7af19556b2e708f2bf350a106687719: updates [Adding features to our bouncing balls demo](https://pr11373.content.dev.mdn.mozit.cloud/en-US/docs/Learn/JavaScript/Objects/Adding_bouncing_balls_features) to use `class` syntax, plus some other minor improvements.

These changes will need a matching PR to https://github.com/mdn/learning-area/tree/main/javascript/oojs, which I will work on now. So I'll keep this draft until that other PR is ready and reviewed. But that doesn't need to block review of this one.

wbamberg commented:
I just opened https://github.com/mdn/learning-area/pull/437 with the corresponding changes to mdn/learning-area. So now I think we are finally ready for a complete review of the PR alongside https://github.com/mdn/learning-area/pull/437.

hamishwillee commented:
@wbamberg The changes look good but I'm not particularly confident that I'm not just skimming over details having already read most of this. In particular, without doing the assignments I think they look sane, but I can't be sure.

So, much improved over last version, definitely not crap, but can't be sure there aren't errors. It would be good if you could get a detail review from someone new.

Rumyra commented:
This all looks good from a read through - the only issues I found were:

- _Test your skills: Object-oriented JavaScript:_ The second `ooJS 2` example isn't showing any example code
- _Test your skills: Object basics:_ `Object 4 basics` is 404'ing

I'm checking on the preview urls so not sure if it's due to that - also there's learning examples to go in (just about to check those) so maybe the latter will resolve with that?

wbamberg commented:
Thank you for looking @Rumyra !

> This all looks good from a read through - the only issues I found were:
> 
>     * _Test your skills: Object-oriented JavaScript:_ The second `ooJS 2` example isn't showing any example code
> 

This is the same as it currently is: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Test_your_skills:_Object-oriented_JavaScript#oojs_2, https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Test_your_skills:_Object-oriented_JavaScript#oojs_3.

The new version here is pretty much like the old version, except I only have 2 tasks (because the old version has task 1 using the old method, then task 2 is "convert to `class` syntax", and in the new version we start out with the `class` syntax).

I agree it is a little jarring to get an empty window here though.

>     * _Test your skills: Object basics:_ `Object 4 basics` is 404'ing
> 
> 
> I'm checking on the preview urls so not sure if it's due to that - also there's learning examples to go in (just about to check those) so maybe the latter will resolve with that?

This should be fixed by the learning-area PR, yes (it's v fiddly updating the learning area because of issues like this, and there's a good chance I will have missed something that we'll only see once it all lands).

wbamberg commented:
> > Re "The weird thing here is that the article says you can't run the example locally because of cross-origin restrictions, but it seems to work just fine if you use an absolute URL? So I changed it. But maybe I'm wrong here?"
> 
> You mean you download https://github.com/mdn/learning-area/tree/main/javascript/oojs/tasks/json and serve locally such that the json and your requesting file are at the same location? That should work.

Yes.

> You might have a problem if you run code in codepen that loads the json from somewhere else though - maybe stuff on github automatically gets shared with the right headers?

I hadn't tried CodePen, but I just did and that works, too. I expect you are right about GitHub. I wonder why the old version didn't think it would work. Maybe it didn't work for some other reason and it was just assumed to be a cross-origin issue?

wbamberg commented:
> > > Re "The weird thing here is that the article says you can't run the example locally because of cross-origin restrictions, but it seems to work just fine if you use an absolute URL? So I changed it. But maybe I'm wrong here?"
> > 
> > 
> > You mean you download https://github.com/mdn/learning-area/tree/main/javascript/oojs/tasks/json and serve locally such that the json and your requesting file are at the same location? That should work.
> 
> Yes.
> 

Um, no, that's not what I meant. I downloaded the example code (HTML+CSS+JS) but it uses an absolute GitHub URL to reference the JSON data. So no, the JSON is not same-origin.

I just took a peek in the network monitor and it looked like https://mdn.github.io/learning-area/javascript/oojs/tasks/json/sample.json is getting served with the `Access-Control-Allow-Origin: *` header, so perhaps that explains it.

hamishwillee commented:
>  https://mdn.github.io/learning-area/javascript/oojs/tasks/json/sample.json is getting served with the Access-Control-Allow-Origin: * header, so perhaps that explains it.

It does.

wbamberg committed:
Merge branch 'main' into js-la-object-oriented-js

Elchi3 reviewed (approved):
I've read through this on Friday and today and it's fantastic work! The diff is surely large and maybe some bug worms are hidden somewhere but I'm confident to approve this! I would say let's land this 🎉

wbamberg commented:
Thanks for the reviews, everyone! @Rumyra , could you take a look at https://github.com/mdn/content/pull/11373#issuecomment-1007618733 and let me know if you're happy to approve this PR or if there's anything else that needs doing?

wbamberg commented:
All right, let's try merging this beast 🤞.
